src/porepy/viz/diagnostics_mixin.py

```diff
@@ -0,0 +1,383 @@
+"""Module for diagnostics of PorePy’s models built on seaborn."""
+
+from __future__ import annotations
+
+import logging
+from collections import defaultdict
+from itertools import product
+from typing import Literal, Optional
+
+import matplotlib
+import numpy as np
+from matplotlib import pyplot as plt
+from scipy.sparse import csr_matrix, spmatrix
+from scipy.sparse.linalg import svds
+
+from porepy import GridLike
+from porepy.grids.mortar_grid import MortarGrid
+from porepy.numerics.ad.equation_system import EquationSystem
+from porepy.numerics.ad.operators import Variable
+
+try:
+    import seaborn as sns  # type: ignore[import]
+except ImportError:
+    _IS_SEABORN_AVAILABLE = False
+else:
+    _IS_SEABORN_AVAILABLE = True
+
+logger = logging.getLogger(__name__)
+
+
+class DiagnosticsMixin:
+    """This is a mixin class for a PorePy model that allows running the model
+    diagnostics.
+
+    Currently supports dividing the matrix into blocks based on different equations and
+    subdomains/interfaces. Can plot:
+    - condition number of each block
+    - absolute maximum value of each block
+    Supports gathering data about one equation/variable block from all subdomains.
+    For detailed information, see the `tutorial <https://github.com/pmgbergen/porepy/blo
+    b/feat-diagnostics/tutorials/diagnostics.ipynb>`__.
+
+    Example:
+        Basic usage:
+
+        >>> from porepy.applications.verification_setups.mandel_biot import MandelSetup
+        >>>
+        >>> class MandelDiagnostics(DiagnosticsMixin, MandelSetup):
+        >>>     pass
+        >>> setup = MandelDiagnostics(params={})
+        >>> porepy.run_time_dependent_model(setup, params={})
+        >>> setup.show_diagnostics()
+
+    """
+
+    submatrices: dict[tuple[int, int], spmatrix]
+    """The submatrices are stored here as they are computed.
+
+    Index represents the pair (row, column) of the block.
+    """
+
+    # The mixin expects the model to have these properties defined.
+    linear_system: tuple[csr_matrix, np.ndarray]
+    equation_system: EquationSystem
+
+    def show_diagnostics(
+        self,
+        grouping: Optional[Literal["subdomains"]] = None,
+        is_plotting_condition_number: bool = False,
+        is_plotting_max: bool = True,
+    ) -> None:
+        """Collects and plots diagnostics from the last assembled Jacobian matrix stored
+        in :attr:`~self.linear_system`.
+
+        It is assumed that the full Jacobian matrix is stored in `self.linear_system`,
+        and full information about the matrix indices is stored in
+        `self.equation_system._equation_image_space_composition`.
+
+        Note:
+            It is assumed that variables with the same name defined on different grids
+            are semantically the same variable.
+
+        Args:
+            grouping (optional): Either "subdomains" or None. If "subdomains",
```

Comment:
How difficult will it be to allow for computing for some subdomains only? I think this would require a second argument (Optional[list[pp.Grid]], None gives all subdomains, empty list give no subdomain). I think this could be useful, what do you say?